keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,859 @@
+one, or n-seconds. Focus is on possibly faster readiness probes, for example every .5 seconds until succss
+and more granular use beyond existing defaults, for example 1.5 second duration instead of picking
+between 1 or 2 second duration.
+Add additional tests cases to the timeout test cases.
+Not breaking backwards compatibility with the V1 API.
+
+
+### Non-Goals
+
+V2 API for existing objects.
+Converting fields from `int32` to `resource.Quantity`.
+
+
+## Proposal
+
+TLDR:
+Add two new optional fields (of type `*int32`) to the Probe struct, which would be used to offset the second based time values in the Probe struct:
+
+- `PeriodMilliseconds`       // How often (in milliseconds) to offset PeriodSeconds when performing the probe (*** to reduce un-necessary resource usage, when periodMilliseconds is used to reduce the period to less than a second the offset is only used until success is reached then no longer).
+- `InitialDelayMilliseconds` // Length of time (in milliseconds) to offset IntialDelaySeconds before health checking is activated.
+
+The seconds and milliseconds fields (as related) will be summed to get the resulting time duration. For example,
+
+```yaml
+...
+#yaml
+periodSeconds: 1
+periodMilliseconds: 500
+...
+```
+
+would be considered a period value of 1.5s or 1500ms. These values are used as `time.Duration` within the prober package ([ex1](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/prober/prober.go#L161) [ex2](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/prober/worker.go#L132) ), so there's no real difference between 1.5s and 1500ms.
+
+A few additional examples:
+
+`periodSeconds: 2` and `periodMilliseconds: -500` would be 1.5s / 1500ms.
+
+`periodSeconds: 1` and `periodMilliseconds: -500` would be 0.5s / 500ms. (*** To reduce un-necessary resource usage, because periodMilliseconds is used to reduce the period to less than a second the offset is only used until success is reached then no longer becoming 1second after success.***)
+
+*** We can reconsider the only go faster until first success policy after alpha (eg. kubelet and/or probe policy for onlyTheFirstTime, always, alwaysWhenWaitingForSucess) ***
+
+More generally, the effective period value would be = `periodSeconds` + `periodMilliseconds`.
+
+There are a few corner cases around default (effective period) values:
+
+`periodSeconds: 0` and `periodMilliseconds: 500` would be 10.5s / 10500ms (as 0 => 10 for `periodSeconds` via [defaults.go](https://github.com/kubernetes/kubernetes/blob/3b13e9445a3bf86c94781c898f224e6690399178/pkg/apis/core/v1/defaults.go#L213-L215))
+
+Each optional millisecond field will be restricted to [-999,999], and the effective sum will never be allowed
+to be less than 0 (if the effective sum is less than 0, it will fail at the validation stage and block deployments).
```

Comment:
Do we really want to allow 1ms probes?

Reply:
no. a min will be selected prob .5sec... see line 178 below